framework/src/utils/PetscSupport.C

```diff
@@ -931,23 +943,75 @@ isSNESVI(FEProblemBase & fe_problem)
 }
 
 void
-setSinglePetscOption(const std::string & name, const std::string & value)
+setSinglePetscOption(const std::string & name,
+                     const std::string & value,
+                     FEProblemBase * const problem /*=nullptr*/)
 {
-  PetscErrorCode ierr;
+  static const TIMPI::Communicator comm_world(PETSC_COMM_WORLD);
+  const TIMPI::Communicator & comm = problem ? problem->comm() : comm_world;
+  LibmeshPetscCall2(comm,
+                    PetscOptionsSetValue(LIBMESH_PETSC_NULLPTR,
+                                         name.c_str(),
+                                         value == "" ? LIBMESH_PETSC_NULLPTR : value.c_str()));
+  const auto lower_case_name = MooseUtils::toLower(name);
+  auto check_problem = [problem, &lower_case_name]()
+  {
+    if (!problem)
+      mooseError("Setting the option '",
+                 lower_case_name,
+                 "' requires passing a 'problem' parameter. Check with your application developer "
+                 "to have them update their code");
+  };
+
+  if (lower_case_name.find("-vec_type") != std::string::npos)
+  {
+    check_problem();
+    for (auto & solver_system : problem->_solver_systems)
+    {
+      auto & lm_sys = solver_system->system();
+      for (auto & [vec_name, vec] : as_range(lm_sys.vectors_begin(), lm_sys.vectors_end()))
+      {
+        libmesh_ignore(vec_name);
+        auto * const petsc_vec = cast_ptr<PetscVector<Number> *>(vec.get());
+        LibmeshPetscCall2(comm, VecSetFromOptions(petsc_vec->vec()));
+      }
+      // The solution vectors aren't included in the system vectors storage
+      auto * petsc_vec = cast_ptr<PetscVector<Number> *>(lm_sys.solution.get());
+      LibmeshPetscCall2(comm, VecSetFromOptions(petsc_vec->vec()));
+      petsc_vec = cast_ptr<PetscVector<Number> *>(lm_sys.current_local_solution.get());
+      LibmeshPetscCall2(comm, VecSetFromOptions(petsc_vec->vec()));
+    }
+  }
+  else if (lower_case_name.find("mat_type") != std::string::npos)
```

Comment:
```suggestion
  // Select matrix type from user-passed PETSc options
  else if (lower_case_name.find("mat_type") != std::string::npos)
```

it is "else if" because lower_case_name only contains a single parameter? checking if I understand this right

Reply:
correct